src/Autonomous.cpp

```diff
@@ -1,227 +1,308 @@
+{
+	poseEstimator.reset(startPose);
+	calibrated = true;
+}
 
-PointXY Autonomous::point_tToPointXY(point_t pnt){
-    return PointXY{static_cast<float>(pnt(0)), static_cast<float>(pnt(1))};
+PointXY Autonomous::point_tToPointXY(const point_t &pnt) const
+{
+	return PointXY{static_cast<float>(pnt(0)), static_cast<float>(pnt(1))};
 }
 
-bool Autonomous::arrived(pose_t gpsPose) {
-    double currX = gpsPose(0);
-    double currY = gpsPose(1);
-    return util::almostEqual(currX, (double) target.x, 0.5) &&
-            util::almostEqual(currY, (double) target.y, 0.5);
+bool Autonomous::arrived(const pose_t &pose) const
+{
+	double currX = pose[0];
+	double currY = pose[1];
+	return util::almostEqual(currX, (double)target.x, 0.5) &&
+		   util::almostEqual(currY, (double)target.y, 0.5);
 }
 
-const std::vector<PointXY> Autonomous::points_tToPointXYs(points_t pnts) {
-  std::vector<PointXY> res;
-  for(point_t &pnt : pnts) {
-    res.push_back(point_tToPointXY(pnt));
-  }
-  return res;
+std::vector<PointXY> Autonomous::points_tToPointXYs(const points_t &pnts) const
+{
+	std::vector<PointXY> res;
+	for (const point_t &pnt : pnts)
+	{
+		res.push_back(point_tToPointXY(pnt));
+	}
+	return res;
 }
 
-double Autonomous::angleToTarget(pose_t gpsPose) {
-    float dy = target.y - (float) gpsPose(1);
-    float dx = target.x - (float) gpsPose(0);
-    double theta = (double) atan2(dy, dx);
-    return theta - gpsPose(2);
+double Autonomous::angleToTarget(const pose_t &gpsPose) const
+{
+	float dy = target.y - (float)gpsPose(1);
+	float dx = target.x - (float)gpsPose(0);
+	double theta = std::atan2(dy, dx);
+	return theta - gpsPose(2);
+}
+
+bool calibratePeriodic(std::vector<pose_t> &poses, const pose_t &pose, pose_t &out)
+{
+	poses.push_back(pose);
+	// 62 samples with 2m std dev and 95% confidence interval gives about +-0.5m
+	if (poses.size() == 62)
+	{
+		pose_t sum;
+		for (const pose_t &p : poses)
+		{
+			sum += p;
+		}
+		sum /= poses.size();
+		out = sum;
+		return true;
+	}
+	else
+	{
+		return false;
+	}
+}
+
+double transformAngle(double currAngle, double targetAngle)
+{
+	double range = 2 * PI;
+	double dist = std::fmod(targetAngle - currAngle, range);
+	double absDist = abs(dist);
+
+	int sign = dist > 0 ? 1 : (dist < 0 ? -1 : 0);
+
+	return currAngle + ((absDist > (range / 2)) ? -sign * (range - absDist) : dist);
+}
+
+double getThetaVel(const PointXY &target, const pose_t &pose, double &thetaErr)
+{
+	double dx = target.x - pose[0];
+	double dy = target.y - pose[1];
+	double targetAngle = atan2(dy, dx);
+	targetAngle = transformAngle(pose[2], targetAngle);
+	thetaErr = targetAngle - pose[2];
+
+	return KP_ANGLE * thetaErr;
+}
+
+double dist(const PointXY &p1, const PointXY &p2)
+{
+	return std::hypot(p1.x - p2.x, p1.y - p2.y);
+}
+
+void filterLandmarks(points_t &landmarks)
+{
+	landmarks.erase(std::remove_if(landmarks.begin(), landmarks.end(),
+								   [](point_t p) { return p[2] == 0; }),
+					landmarks.end());
```

Comment:
We might not want to do this, because the index in the array indicates which landmark it is. So e.g. if only one of the landmarks is visible, we'll erase all the others and then we won't know which landmark we're looking at.

Reply:
Oh, I didn't realize that's how the landmarks worked, I'll fix that. I assume the landmark for the first leg is at index 0, right? Is there one landmark per leg?

Reply:
No, some legs have gates instead of posts, which means there are two landmarks instead of one. Best just to think about landmarks as AR tags, and the mapping from legs to AR tags is done using the `URCLeg` objects and the `getLeg()` method.

Reply:
But anyway, doing this properly would require passing in the URCLeg object instead of just a PointXY, which is exactly what #51 does. So we'll deal with it in #51 after this is merged. The updated code looks good.